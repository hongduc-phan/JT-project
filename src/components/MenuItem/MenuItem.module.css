.wrap {
  display: flex;
  cursor: pointer;
  padding: 0.75rem 1rem;
  transition: background-color 150ms ease-in-out;
}

.wrap:hover,
.selected {
  background-color: var(--grey-offwhite);
}
