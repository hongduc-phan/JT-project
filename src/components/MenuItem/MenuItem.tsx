import React, {AllHTMLAttributes, FunctionComponent, ReactNode} from 'react';
import clsx from 'clsx';
import Typo, {TypoColors} from '../Typo';

import styles from './MenuItem.module.css';

export interface IMenuItemProps extends AllHTMLAttributes<HTMLLIElement> {
  value: any;
  color?: TypoColors;
  children: ReactNode;
  selected?: boolean;
}

const MenuItem: FunctionComponent<IMenuItemProps> = ({
  value,
  children,
  className,
  color,
  selected,
  ...other
}: IMenuItemProps) => {
  return (
    <Typo<HTMLLIElement>
      tag="li"
      className={clsx(styles.wrap, className, selected && styles.selected)}
      color={color}
      {...other}
    >
      {children}
    </Typo>
  );
};

export default MenuItem;
