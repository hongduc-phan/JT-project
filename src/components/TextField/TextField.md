Textbox

```js
const TextField = require('./').default;

<TextField label="Username" />;
```

Textbox with Assistive text

```js
const TextField = require('./').default;

<TextField label="Username" assistiveText="Assistive text" />;
```

Textbox disabled

```js
const TextField = require('./').default;

<TextField disabled={true} label="Username" assistiveText="Assistive text" />;
```

Tall textbox

```js
const TextField = require('./').default;

<TextField label="Username" display="tall" />;
```

Error textbox

```js
const TextField = require('./').default;

<TextField label="Username" assistiveText="Invalid username" error={true} />;
```

Select box

```js
const TextField = require('./').default;
const MenuItem = require('../MenuItem').default;

<TextField label="Country" select={true} value="nKorea">
  <MenuItem value="usa">USA</MenuItem>
  <MenuItem value="russia">Russia</MenuItem>
  <MenuItem value="nKorea">North Korea</MenuItem>
  <MenuItem value="china">China</MenuItem>
</TextField>;
```

Select box disabled

```js
const TextField = require('./').default;
const MenuItem = require('../MenuItem').default;

<TextField label="Country" disabled={true} select={true} value="nKorea">
  <MenuItem value="usa">USA</MenuItem>
  <MenuItem value="russia">Russia</MenuItem>
  <MenuItem value="nKorea">North Korea</MenuItem>
  <MenuItem value="china">China</MenuItem>
</TextField>;
```
