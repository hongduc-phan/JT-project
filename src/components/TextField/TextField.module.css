/* Foundation */

.rootDisabled {
  opacity: 0.5;
}

.textField {
  position: relative;
  border-radius: 0.1875rem;
  border: 0.09375rem solid var(--grey-light);
}

.textField:focus {
  display: none;
}

.textField.error {
  border-color: var(--red-negative) !important;
}

.textField:hover {
  border-color: var(--black);
}

.textField:hover .label {
  color: var(--black);
}

/* Label */

.formLabel {
  position: relative;
  flex-grow: 1;
}

.label {
  position: absolute;
  top: 0;
  left: 0;
  font-size: 1rem;
  color: var(--grey-light);
  background-color: var(--white);
  transform: translate(0, -0.125rem) scale(1);
  transition: color 200ms var(--decelerate-easing),
    transform 200ms var(--decelerate-easing) 0ms;
  pointer-events: none;
}

.label.error {
  color: var(--red-negative) !important;
}

.activedLabel {
  padding: 0 0.25rem;
  color: var(--black);
  transform: translate(-0.125rem, -1.125rem) scale(0.75);
  transform-origin: top left;
}

.tall .activedLabel {
  transform: translate(-0.125rem, -1.625rem) scale(0.75);
}

/* Input */

.input {
  width: 100%;
  font-size: 1rem;
  line-height: 1.5;
  color: var(--black);
  outline: none;
  border: none;
}

.inputReadOnly {
  cursor: not-allowed;
}

::placeholder {
  color: var(--grey-medium);
}

.input:disabled {
  background-color: transparent;
}

.icon {
  display: inline-flex;
  color: var(--grey-light);
  font-size: 1.5rem;
  margin-left: 1rem;
}

/* Group */

.formGroup {
  display: flex;
  align-items: center;
}

.formGroup.reverse {
  flex-direction: row-reverse;
}

.formGroup.reverse .icon {
  margin-left: 0;
  margin-right: 0.6875rem;
}

/* Assistive Text */

.assistiveText {
  padding: 0 0.8125rem;
  margin-top: 0.25rem;
}

/* Activation */

.normal {
  padding: 0.5rem 0.8125rem;
}

.active {
  color: var(--black);
}

.tall {
  padding: 1rem 0.8125rem;
}

.focus {
  border-color: var(--blue-default) !important;
}

.arrow {
  z-index: 1;
  top: 50%;
  margin-top: -0.5rem;
  position: absolute;
  right: 0.8125rem;
  color: var(--grey-light);
}

.inputSelect {
  cursor: pointer;
  min-height: 1.125rem;
}
