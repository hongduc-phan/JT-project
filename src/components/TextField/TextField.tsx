import React, {
  AllHTMLAttributes,
  useState,
  useRef,
  ReactNode,
  useEffect,
} from 'react';
import clsx from 'clsx';

import {TextFieldAlignment} from './types';

import Typo, {TypoVariants, TypoColors} from '../Typo';
import {ArrowDown} from '../Icons';
import Select from '../Select';

import styles from './TextField.module.css';

export interface ITextFieldProps extends AllHTMLAttributes<HTMLInputElement> {
  active?: boolean;
  disabled?: boolean;
  label?: string;
  icon?: ReactNode;
  assistiveText?: string;
  error?: boolean;
  display?: string;
  alignIcon?: TextFieldAlignment;
  select?: boolean;
  children?: ReactNode;
  /** Callback when change selected value on select box */
  onChangeSelectValue?: (value: any) => void;
  /** Set init open select box */
  open?: boolean;
  inputRef?: React.RefObject<HTMLInputElement>;
  className?: string;
  classes?: {
    assistiveText?: string;
    textField?: string;
    textFieldFocus?: string;
    textFieldError?: string;
    arrow?: string;
    inputSelect?: string;
  };
}

function getDisplay(display: string) {
  if ('tall' === display) {
    return styles.tall;
  }
  return styles.normal;
}

function getColor(error: boolean, active: boolean): TypoColors {
  if (error) {
    return TypoColors.red;
  }
  return active ? TypoColors.black : TypoColors.greyMedium;
}

const TextField = ({
  required,
  name,
  children,
  select,
  label,
  icon,
  assistiveText = '',
  error = false,
  display = 'normal',
  alignIcon = TextFieldAlignment.right,
  onBlur,
  onFocus,
  value,
  defaultValue,
  readOnly,
  onChangeSelectValue,
  open,
  disabled,
  onChange,
  inputRef,
  hidden,
  className,
  active = false,
  classes = {},
  ...others
}: ITextFieldProps) => {
  const inputEl = useRef(null);
  const [isActive, setActive] = useState(active || Boolean(value));
  const [focus, setFocus] = useState(false);
  const [openSelect, setOpenSelect] = useState(open);
  const [valueSelect, setValueSelect] = useState(value);
  const [displaySelectValue, setDisplaySelectValue] = useState('');

  function handlerOnFocus(e: React.FocusEvent<HTMLInputElement>) {
    if (!active) {
      setActive(true);
      setFocus(true);
    }
    e.persist();
    if (onFocus) {
      onFocus(e);
    }
  }

  function onFocusOut(e: React.FocusEvent<HTMLInputElement>) {
    if (!active) {
      setActive(Boolean(e.currentTarget.value) || false);
      setFocus(false);
    }
    e.persist();
    if (onBlur) {
      onBlur(e);
    }
  }

  function onClickParent() {
    setOpenSelect(!openSelect);
    setFocus(true);
  }

  function onRequestCloseSelect() {
    setOpenSelect(false);
    setFocus(false);
  }

  function handlerOnChangeValueSelect(v: any) {
    setValueSelect(v);
    if (onChangeSelectValue) {
      onChangeSelectValue(v);
    }
    if (onChange) {
      // Manually call onchange. So react-hoc-form-validatable can capture new values
      onChange({
        target: {
          name,
          value: v,
        },
        persist(): void {
          //
        },
      } as any);
    }
  }

  useEffect(() => {
    if (!active) {
      if (value) {
        setActive(true);
      } else {
        setActive(false);
      }
    }

    if (select) {
      setValueSelect(value);
    }
  }, [value]);

  return (
    <div
      className={clsx(styles.root, disabled && styles.rootDisabled, className)}
      hidden={hidden}
    >
      <div
        ref={inputEl}
        onClick={select && !disabled ? onClickParent : undefined}
        className={clsx(
          styles.textField,
          getDisplay(display),
          {
            [styles.active]: isActive,
            [styles.focus]: focus,
            [styles.error]: error,
            [classes.textFieldFocus as string]: isActive,
            [classes.textFieldError as string]: error,
          },
          classes.textField,
        )}
      >
        {select && (
          <span className={clsx(styles.arrow, classes.arrow)}>
            <ArrowDown />
          </span>
        )}
        <div
          className={clsx(styles.formGroup, {
            [styles.reverse]: 'left' === alignIcon,
          })}
        >
          <div className={styles.formLabel}>
            {label && (
              <Typo
                className={clsx(styles.label, {
                  [styles.activedLabel]: select
                    ? Boolean(valueSelect)
                    : isActive,
                  [styles.error]: error,
                })}
                variant={TypoVariants.subTitle}
              >
                {label}
                {required && '*'}
              </Typo>
            )}
            <input
              {...others}
              className={clsx(styles.input, readOnly && styles.inputReadOnly)}
              onFocus={!readOnly ? handlerOnFocus : undefined}
              onBlur={!readOnly ? onFocusOut : undefined}
              value={select ? valueSelect : value}
              readOnly={select ? true : readOnly}
              disabled={disabled}
              hidden={select}
              onChange={onChange}
              ref={inputRef}
              name={name}
              required={required}
              defaultValue={defaultValue}
            />
            {select && (
              <div
                className={clsx(
                  styles.input,
                  select && styles.inputSelect,
                  classes.inputSelect,
                )}
              >
                {displaySelectValue}
              </div>
            )}
          </div>
          {icon && <span className={styles.icon}>{icon}</span>}
        </div>
      </div>
      {assistiveText && (
        <Typo
          tag="div"
          variant={TypoVariants.subTitle}
          color={getColor(error, isActive)}
          className={clsx(styles.assistiveText, classes.assistiveText)}
        >
          {assistiveText}
        </Typo>
      )}
      {select && (
        <Select
          value={valueSelect}
          onChangeValue={handlerOnChangeValueSelect}
          onChangeDisplayValue={setDisplaySelectValue}
          onRequestCloseSelect={onRequestCloseSelect}
          open={openSelect}
          anchor={inputEl.current}
        >
          {children}
        </Select>
      )}
    </div>
  );
};

export default TextField;
