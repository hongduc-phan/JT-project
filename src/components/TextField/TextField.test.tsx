import React from 'react';
import renderer from 'react-test-renderer';
import {mount, ReactWrapper} from 'enzyme';
import {spy} from 'sinon';

import TextField from './TextField';
import MenuItem from '../MenuItem';

describe('<TextField />', () => {
  let textField: any;
  let component: ReactWrapper;

  beforeEach(() => {
    textField = <TextField label="Label" />;

    component = mount(textField);
  });

  it("Shouldn't display icon", () => {
    const imgs = component.find('img');
    expect(imgs.length).toEqual(0);
  });

  it('Should display icon', () => {
    textField = (
      <TextField
        label="Label"
        icon="https://cdn4.iconfinder.com/data/icons/ionicons/512/icon-ios7-eye-256.png"
      />
    );
    const imgs = component.find('img');
    expect(imgs.length).toEqual(0);
  });

  it("Shouldn't focus on TextField", () => {
    let classNames;

    classNames = component.find('.label.activedLabel');
    expect(classNames.length).toEqual(0);
  });

  it('Should focus on Textfield', () => {
    component.find('input').simulate('focus');
    expect(
      component
        .find('div')
        .at(3)
        .find('span')
        .hasClass('activedLabel'),
    ).toEqual(true);
  });

  it('Should match textField snapshot', () => {
    const tree = renderer.create(textField).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should match TextField select snapshot', () => {
    const tree = renderer
      .create(
        <TextField label="Country" select={true} value="nKorea">
          <MenuItem value="usa">USA</MenuItem>
          <MenuItem value="russia">Russia</MenuItem>
          <MenuItem value="nKorea">North Korea</MenuItem>
          <MenuItem value="china">China</MenuItem>
        </TextField>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('TextField type select should work', () => {
    const cbChangeSelect = spy();

    const wrapper = mount(
      <TextField
        label="Country"
        select={true}
        onChangeSelectValue={cbChangeSelect}
        value="usa"
      >
        <MenuItem value="usa">USA</MenuItem>
        <MenuItem value="russia">Russia</MenuItem>
        <MenuItem value="nKorea">North Korea</MenuItem>
        <MenuItem value="china">China</MenuItem>
      </TextField>,
    );

    wrapper.find('input').simulate('click');
    expect(
      wrapper
        .find('li')
        .at(0)
        .text(),
    ).toEqual('USA');
    expect(wrapper.find('input').props().value).toEqual('usa');
    expect(wrapper.find('.inputSelect').text()).toEqual('USA');
    wrapper
      .find('li')
      .at(1)
      .simulate('click');
    expect(wrapper.find('input').props().value).toEqual('russia');
    expect(wrapper.find('.inputSelect').text()).toEqual('Russia');
    expect(cbChangeSelect.calledWith('russia')).toBeTruthy();
  });
});
