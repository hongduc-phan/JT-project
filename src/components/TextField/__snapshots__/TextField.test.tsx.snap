// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Should match TextField select snapshot 1`] = `
<div
  className="root"
>
  <div
    className="textField normal active"
    onClick={[Function]}
  >
    <span
      className="arrow"
    >
      <svg
        height="1em"
        viewBox="0 0 20 20"
        width="1em"
      >
        <g
          fill="none"
          fillRule="evenodd"
        >
          <path
            d="M6.175 6.667L10 10.483l3.825-3.816L15 7.842l-5 5-5-5z"
            fill="currentColor"
            fillRule="nonzero"
          />
          <path
            d="M0 20h20V0H0z"
          />
        </g>
      </svg>
    </span>
    <div
      className="formGroup"
    >
      <div
        className="formLabel"
      >
        <span
          className="typo label activedLabel subTitle black left"
        >
          Country
        </span>
        <input
          className="input"
          hidden={true}
          onBlur={[Function]}
          onFocus={[Function]}
          readOnly={true}
          value="nKorea"
        />
        <div
          className="input inputSelect"
        >
          
        </div>
      </div>
    </div>
  </div>
  
</div>
`;

exports[`<TextField /> Should match textField snapshot 1`] = `
<div
  className="root"
>
  <div
    className="textField normal"
  >
    <div
      className="formGroup"
    >
      <div
        className="formLabel"
      >
        <span
          className="typo label subTitle black left"
        >
          Label
        </span>
        <input
          className="input"
          onBlur={[Function]}
          onFocus={[Function]}
        />
      </div>
    </div>
  </div>
  
</div>
`;
